src/types.rs

```diff
@@ -46,64 +50,138 @@ pub struct LookupEntry {
+        Ok(Self {
             envelope_type: bytes[0],
-            iv: bytes[1..13].try_into().unwrap(),
+            iv: bytes[1..13].try_into()?,
             sealbox: bytes[13..].to_vec(),
-        }
+            opts: EnvelopeType0 {},
+        })
     }
 }
 
-impl Envelope for EnvelopeType1 {
-    fn to_bytes(&self) -> Vec<u8> {
+impl Envelope<EnvelopeType1> {
+    pub fn new(encryption_key: &str, data: impl Serialize, pubkey: [u8; 32]) -> Result<Self> {
+        let serialized = serde_json::to_vec(&data)?;
+        let iv = generate_nonce().into();
+        let encryption_key = hex::decode(encryption_key)?;
+
+        let cipher = ChaCha20Poly1305::new(GenericArray::from_slice(&encryption_key));
+
+        let sealbox = cipher
+            .encrypt(&iv, serialized.as_bytes())
+            .map_err(|_| crate::error::Error::EncryptionError("Encryption failed".into()))?;
+
+        Ok(Self {
+            envelope_type: 1,
+            opts: EnvelopeType1 { pubkey },
+            iv: iv.into(),
+            sealbox,
+        })
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
         let mut serialized = vec![];
         serialized.push(self.envelope_type);
-        serialized.extend_from_slice(&self.pubkey);
+        serialized.extend_from_slice(&self.opts.pubkey);
         serialized.extend_from_slice(&self.iv);
         serialized.extend_from_slice(&self.sealbox);
         serialized
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
-        Self {
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self> {
+        Ok(Self {
             envelope_type: bytes[0],
-            pubkey: bytes[1..33].try_into().unwrap(),
-            iv: bytes[33..45].try_into().unwrap(),
+            opts: EnvelopeType1 {
+                pubkey: bytes[1..33].try_into()?,
+            },
+            iv: bytes[33..45].try_into()?,
             sealbox: bytes[45..].to_vec(),
-        }
+        })
+    }
+
+    pub fn pubkey(&self) -> [u8; 32] {
+        self.opts.pubkey
     }
 }
+
+pub trait EnvelopeType {}
+
+#[derive(Serialize)]
+pub struct EnvelopeType0 {}
+impl EnvelopeType for EnvelopeType0 {}
+
+pub struct EnvelopeType1 {
+    pub pubkey: [u8; 32],
+}
+
+impl EnvelopeType for EnvelopeType1 {}
+
+fn generate_nonce() -> GenericArray<u8, U12> {
+    let uniform = Uniform::from(0u8..=255);
+
+    let mut rng = OsRng {};
+    GenericArray::from_iter(uniform.sample_iter(&mut rng).take(12))
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Subscribtion {
+    pub topic: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Unsubscribe {
+    pub topic: String,
+    pub id: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct Notification {
+    pub title: String,
+    pub body: String,
+    pub icon: String,
+    pub url: String,
+    #[serde(rename = "type")]
+    #[serde(default = "default_notification_type")]
+    pub notification_type: String,
```

Comment:
You can use raw identifier: `r#type`. Or just call it `ty`, it's pretty common alias for type